src/signature.c

```diff
@@ -24,198 +24,364 @@
- *   FORGIVE .. do everything, always return success
- *   REAL ..... do everything, return the real status
- */
-#define IMPL_FAKE 0
-#define IMPL_FORGIVE 1
-#define IMPL_REAL 2
+#define BUFFER_SIZE   4096
 
-#warning "TODO pick signing scheme"
-#if defined(SWUPD_LINUX_ROOTFS)
-#define IMPL IMPL_FAKE
+#ifdef SIGNATURES
+       bool verify_signatures = true;
+#else
+       bool verify_signatures = false;
 #endif
 
-#if IMPL != IMPL_FAKE
-
-static X509_STORE *create_store(const char *, const char *, const char *);
-
-static char *VERIF_FAIL = "Signature verification failed";
-static char *XSTORE_FAIL = "XSTORE creation failed";
+static bool validate_signature(FILE *, FILE *);
+static bool validate_certificate(void);
+static int verify_callback(int, X509_STORE_CTX *);
+static bool get_pubkey();
+static bool get_certificates(void);
 
 static bool initialized = false;
+static EVP_PKEY *pkey;
+static X509 *cert;
+static char *chain;
+static char *crl;
+static char *certname;
+static char *ca_dirname;
 
-static X509_STORE *x509_store = NULL;
-
-bool signature_initialize(const char *ca_cert_filename)
+/* This function must be called before trying to sign any file.
+ * It loads string for errors, and ciphers are auto-loaded by OpenSSL now.
+ * If this function fails it may be because the certificate cannot
+ * be validated.
+ *
+ * ca_cert_filename: is the swupd certificate that contains the public key
+ * for signature verification.
+ *
+ * returns: true if can initialize and validate certificates, otherwise false */
+bool initialize_signature(void)
 {
 	if (initialized) {
 		return true;
 	}
-	OpenSSL_add_all_algorithms();
+
 	ERR_load_crypto_strings();
-	x509_store = create_store(ca_cert_filename, NULL, NULL);
-	if (x509_store == NULL) {
-		ERR_free_strings(); // undoes ERR_load_crypto_strings
-		EVP_cleanup();      // undoes OpenSSL_add_all_algorithms
-		return false || (IMPL == IMPL_FORGIVE);
+
+	if (!get_certificates()) {
+		goto fail;
+	}
+
+	if (!validate_certificate()) {
+		goto fail;
 	}
+
+	if (!get_pubkey()) {
+		goto fail;
+	}
+
 	initialized = true;
 	return true;
+fail:
+	ERR_free_strings();
+	EVP_cleanup();
+	return false;
 }
 
-void signature_terminate(void)
+/* Delete the memory used for string errors as well as memory allocated for
+ * certificates and private keys. */
+void terminate_signature(void)
 {
 	if (initialized) {
-		X509_STORE_free(x509_store); // undocumented...
-		ERR_free_strings();	  // undoes ERR_load_crypto_strings
-		EVP_cleanup();		     // undoes OpenSSL_add_all_algorithms
+		X509_free(cert);
+		ERR_free_strings();
+		EVP_PKEY_free(pkey);
+		EVP_cleanup();
 		initialized = false;
 	}
 }
 
-bool signature_verify(const char *data_filename, const char *sig_filename)
+/* Get the paths and filenames of certificates */
+bool get_certificates(void)
 {
-	BIO *bio_data = NULL;
-	BIO *bio_sig = NULL;
-	PKCS7 *pkcs7 = NULL;
-	int ret;
+	certname = (char *) X509_get_default_cert_file();
```

Comment:
We're going to be using one and only one specific cert for swupd signing, so instead of loading the full system-wide compliment of certs, I suggest we remove this call and replace it a call loading just the cert we want.  

X509_load_cert_file() may do the trick